optimum/habana/transformers/generation/utils.py

```diff
@@ -1239,8 +1243,426 @@ def contrastive_search(
+                if token_idx is not None:
+                    # Reset inputs to top_k_ids for static shapes to override selection by token_idx
+                    if self.config.is_encoder_decoder:
+                        next_model_inputs["decoder_input_ids"] = top_k_ids.view(-1, 1)
+                    else:
+                        next_model_inputs["input_ids"] = top_k_ids.view(-1, 1)
+
+                outputs = self(
+                    **next_model_inputs,
+                    return_dict=True,
+                    output_hidden_states=True,
+                    output_attentions=output_attentions,
+                )
+
+            # name is different for encoder-decoder and decoder-only models
+            if self.config.is_encoder_decoder:
+                next_hidden = outputs.decoder_hidden_states[-1]
+                full_hidden_states = outputs.decoder_hidden_states
+            else:
+                next_hidden = outputs.hidden_states[-1]
+                full_hidden_states = outputs.hidden_states
+
+            logits = outputs.logits[:, -1, :]
+
+            context_hidden = last_hidden_states.repeat_interleave(top_k, dim=0)
+
+            # compute the degeneration penalty and re-rank the candidates based on the degeneration penalty and the
+            # model confidence. Keeping `selected_idx` on CPU enables multi-device contrastive search and doesn't
+            # introduce (noticeable) slowdowns on single-device runs.
+            selected_idx = _ranking_fast(context_hidden, next_hidden, top_k_probs, penalty_alpha, top_k)
+            selected_idx = selected_idx.to("cpu")
+
+            # prepare for the next step: (1) next token_id; (2) past_key_values; (3) last_hidden_states for computing
+            # the degeneration penalty; (4) logits for selecting next top-k candidates; (5) selected tokens scores
+            # (model confidence minus degeneration penalty); (6) decoder hidden_states
+            next_tokens = top_k_ids[range(len(top_k_ids)), selected_idx]
+            next_hidden = torch.stack(torch.split(next_hidden.squeeze(dim=1), top_k))
+            next_hidden = next_hidden[range(batch_size), selected_idx, :]
+            last_hidden_states = torch.cat([last_hidden_states, next_hidden.unsqueeze(1)], dim=1)
+
+            next_decoder_hidden_states = ()
+            for layer in full_hidden_states:
+                layer = torch.stack(torch.split(layer, top_k))[range(batch_size), selected_idx, :]
+                next_decoder_hidden_states += (layer,)
+
+            # generate past_key_values cache of only the selected token
+            if sequential:
+                next_model_input = self.prepare_inputs_for_generation(
+                    top_k_ids[:, selected_idx].view(-1, 1), **model_kwargs
+                )
+                if token_idx is not None:
+                    # Reset inputs to top_k_ids for static shapes to override selection by token_idx
```

Comment:
@regisss @jiminha The shape of `top_k_ids` is different from the search input so `prepare_inputs_for_generation` for static shapes returns zeros because `token_idx` is greater than the dimensions of  `top_k_ids`. 

Alternative ways to fix this would be to temporarily set  `token_idx=1` before preparing inputs, or to create a padded  `top_k_ids` for max_length which we can update. Please let me know what approach to use.

Reply:
@libinta  @ssarkar2  could you comment on Soila's question above and review the code?

Reply:
@skavulya  we need to pad the top_k_ids for static shape

Reply:
@libinta  I padded the top_k_ids